optuna/integration/allennlp/_train.py

```diff
@@ -0,0 +1,378 @@
+        # passed the wrong thing - cuda_devices are required.
+        device_ids = distributed_params.pop("cuda_devices", None)
+        multi_device = isinstance(device_ids, list) and len(device_ids) > 1
+        num_nodes = distributed_params.pop("num_nodes", 1)
+
+        if not (multi_device or num_nodes > 1):
+            raise ConfigurationError(
+                "Multiple cuda devices/nodes need to be configured to run distributed training."
+            )
+        check_for_gpu(device_ids)
+
+        primary_addr = distributed_params.pop("primary_address", "127.0.0.1")
+        if primary_addr in ("127.0.0.1", "0.0.0.0", "localhost"):
+            # If running locally, we can automatically find an open port if one is not specified.
+            primary_port = (
+                distributed_params.pop("primary_port", None) or common_util.find_open_port()
+            )
+        else:
+            # Otherwise we require that the port be specified.
+            primary_port = distributed_params.pop("primary_port")
+
+        num_procs = len(device_ids)
+        world_size = num_nodes * num_procs
+
+        # Creating `Vocabulary` objects from workers could be problematic since
+        # the data loaders in each worker will yield only `rank` specific
+        # instances. Hence it is safe to construct the vocabulary and write it
+        # to disk before initializing the distributed context. The workers will
+        # load the vocabulary from the path specified.
+        vocab_dir = os.path.join(serialization_dir, "vocabulary")
+        if recover:
+            vocab = Vocabulary.from_files(vocab_dir)
+        else:
+            vocab = training_util.make_vocab_from_params(
+                params.duplicate(), serialization_dir, print_statistics=dry_run
+            )
+        params["vocabulary"] = {
+            "type": "from_files",
+            "directory": vocab_dir,
+            "padding_token": vocab._padding_token,
+            "oov_token": vocab._oov_token,
+        }
+
+        logging.info(
+            "Switching to distributed training mode since multiple GPUs are configured | "
+            f"Primary is at: {primary_addr}:{primary_port} | Rank of this node: {node_rank} | "
+            f"Number of workers in this node: {num_procs} | Number of nodes: {num_nodes} | "
+            f"World size: {world_size}"
+        )
+
+        assert trial is not None
+
+        if isinstance(trial.study._storage, InMemoryStorage):
+            message = "InMemoryStorage is not supported in distributed configuration."
+            message += " You have to use RDB or Redis to use Optuna with AllenNLP distributed."
+            raise ValueError(message)
+
+        elif isinstance(trial.study._storage, _CachedStorage):
+            # Reconstruct storage to purge a old cache.
+            if isinstance(trial.study._storage, _CachedStorage):
+                trial.study._storage = _CachedStorage(trial.study._storage._backend)
```

Comment:
I don't know if I can explain it well, but I try to do.

![trial_distributed](https://user-images.githubusercontent.com/5164000/125449830-3a89e65d-3cc1-4b0f-af69-7ab2b990ab8c.png)

Worker processes are spawned from the main process. Each worker runs a training loop and reports a metric to Optuna. Optuna's trial object is copied to each worker.

When `trial.report` is invoked, an objective value is recorded to a storage but Optuna's `CachedStorage` mechanism caches a state in memory and avoid executing a query to the storage. Each worker terminates after training thus the changes in a worker's state are not propagated to the main process.

To get previous trials' information, I re-create a storage object (it fetches information from a remote storage) before starting training.